### PR TITLE
Reg desc anno

### PR DESCRIPTION
**Pull request by depappas-sifive**
This code need to be cleaned up.

### CODE REVIEW

**Comment by mwachs5** [comment]
Also, can we have better commit messages in the final version than multiple commits that just say `annos`? I assume those will be squashed into something more meaningful?

File: src/main/scala/regmapper/Annotation.scala [hunk 1]
```diff
@@ -2,17 +2,121 @@
...
+import firrtl.annotations._
+import freechips.rocketchip.regmapper.RegFieldAccessType.RegFieldAccessType
+import freechips.rocketchip.regmapper.RegFieldRdAction.RegFieldRdAction
+import freechips.rocketchip.regmapper.RegFieldWrType.RegFieldWrType
+import org.json4s.JsonAST.JValue
 import org.json4s.JsonDSL._
 import org.json4s.jackson.JsonMethods.{pretty, render}
 
+// Do not call this function
+// This function is called by DescribedRegChiselAnnotation
```
Comment (mwachs5) [inline src/main/scala/regmapper/Annotation.scala]: then can't it be placed inside that class and marked `private`?

File: src/main/scala/regmapper/Annotation.scala [hunk 2]
```diff
@@ -2,17 +2,121 @@
...
+    val regDesc: freechips.rocketchip.regmapper.RegFieldDesc = regField.desc.get
+
+    //require("", regDesc.desc != None)
+
+    ( ("byteOffset"   -> s"0x${byteOffset.toHexString}") ~
+      ("bitOffset"    -> bitOffset) ~
+      ("bitWidth"     -> regField.width) ~
+      ("name"         -> regDesc.name) ~
+      ("description"  -> regDesc.desc) ~
+// TODO      ("resetValue"   -> regDesc.reset.getOrElse("0")) ~
```
Comment (mwachs5) [inline src/main/scala/regmapper/Annotation.scala]: why can't this handle optional parameters?

File: src/main/scala/regmapper/Annotation.scala [hunk 3]
```diff
@@ -21,3 +125,64 @@ object RegMappingAnnotation {
...
+case class RegFieldDescSer(
+  name: String,
+  desc: String,
+  group: String,
+  groupDesc: String,
+  access: String,
+  wrType: String,
+  rdAction: String,
+  volatile: Boolean,
+  reset: BigInt
```
Comment (mwachs5) [inline src/main/scala/regmapper/Annotation.scala]: not all registers have reset values. This is a pretty critical thing to capture if it's not reset.

File: src/main/scala/regmapper/DescribedReg.scala [hunk 4]
```diff
@@ -23,7 +24,14 @@ object DescribedReg {
     enumerations: Map[BigInt, (String, String)] = Map()): (T, RegFieldDesc) = {
     val rdesc = RegFieldDesc(name, desc, None, None,
       access, wrType, rdAction, volatile, reset.map{_.litValue}, enumerations)
+
     val reg = reset.map{i => RegInit(i)}.getOrElse(Reg(gen))
+    val byte = 0 // TODO
```
Comment (mwachs5) [inline src/main/scala/regmapper/DescribedReg.scala]: why is this in here? The register description doesn't know or need to know about byte and offset.

Comment (mwachs5) [reply src/main/scala/regmapper/DescribedReg.scala]: a `DescribedReg` may not even end up in a regmap call ever... I don't get what this is here for.

File: src/main/scala/regmapper/RegField.scala [hunk 5]
```diff
@@ -141,9 +142,29 @@ case class RegField(width: Int, read: RegReadFn, write: RegWriteFn, desc: Option
 {
   require (width > 0, s"RegField width must be > 0, not $width")
 
-  def pipelined = !read.combinational || !write.combinational
+  def pipelined: Boolean = !read.combinational || !write.combinational
```
Comment (mwachs5) [inline src/main/scala/regmapper/RegField.scala]: why are we randomly adding the type to some functions/vars and not others? This seems unrelated to the PR

File: src/main/scala/regmapper/Annotation.scala (hunk 1 above)
Comment (depappas-sifive) [reply src/main/scala/regmapper/Annotation.scala]: Ask Jack

File: src/main/scala/regmapper/RegField.scala (hunk 5 above)
Comment (depappas-sifive) [reply src/main/scala/regmapper/RegField.scala]: Rocket-chip is not run against the lint rules and does not pass lint. 

Public functions without a return type are a lint violation.

The return types were not added randomly. They were added to functions that were being modified. 

The Intellij IDE considers lint violations to be errors and this causes problems when using the IDE.

Henry suggested making a separate PR for rocket-chip lint fixes.

I can remove the Lint fixes and put them in a PR.

Comment (mwachs5) [reply src/main/scala/regmapper/RegField.scala]: Sounds good... could address this ancient issue: https://github.com/freechipsproject/rocket-chip/issues/255

Comment (richardxia) [reply src/main/scala/regmapper/RegField.scala]: I don't think we should make lint changes *just* based on what your editor considers a linting issue. If we don't unilaterally enforce lint at the CI level, then anyone not using the same editor as you may just continue to commit code that your editor considers to be a lint violation. If you want to make style changes, then please add a CI check to lint the code.

File: src/main/scala/regmapper/Annotation.scala (hunk 1 above)
Comment (depappas-sifive) [reply src/main/scala/regmapper/Annotation.scala]: Comment updated

File: src/main/scala/regmapper/Annotation.scala (hunk 2 above)
Comment (depappas-sifive) [reply src/main/scala/regmapper/Annotation.scala]: Optional params are added but I am not sure what I should put for the default param values in the Else.

Comments?

File: src/main/scala/regmapper/Annotation.scala (hunk 3 above)
Comment (depappas-sifive) [reply src/main/scala/regmapper/Annotation.scala]: Should I add a hasRest boolean?

File: src/main/scala/regmapper/Annotation.scala (hunk 2 above)
Comment (mwachs5) [reply src/main/scala/regmapper/Annotation.scala]: I don't understand why you need anythng if it's optional. If it's not there, then there IS NO reset value. I don't know what you are doing with this so what the limitation is. If this is emitting JSON as it was before, then if the resetValue is omitted from the JSON, then the field is not reset.

File: src/main/scala/regmapper/Annotation.scala (hunk 3 above)
Comment (mwachs5) [reply src/main/scala/regmapper/Annotation.scala]: Again, I don't know what you are doing and why the absence of a `resetValue` field in the emitted JSON is not allowed / sufficient. If there is no `resetValue`, that is the same as `hasReset = false`

File: src/main/scala/regmapper/RegField.scala (hunk 5 above)
Comment (richardxia) [reply src/main/scala/regmapper/RegField.scala]: There is no scalastyle-config.xml in rocket-chip.

File: src/main/scala/regmapper/Annotation.scala [hunk 6]
```diff
@@ -2,22 +2,196 @@
...
+object RegFieldDescAnnotation {
+  def nameMap(state: CircuitState): Map[String,Seq[String]] = {
+    state.annotations.collect {
+      case RegFieldDescAnnotation(mname, cname) => mname.toString -> cname
+    }.toMap
+  }
+}
+
+
+case class DescribedRegChiselAnnotation(
```
Comment (mwachs5) [inline src/main/scala/regmapper/Annotation.scala]: I think we chatted about this, but there is no need for a `DescribedReg` annotation. `DescribedReg` is just a convenience for the designer in coding up a `RegFieldDesc`. It is only once a `RegFieldDesc` is applied to a `RegField` that is passed into a call to `regmap()` that we care about it. At the point of the `DescribedReg` call it is meaningless.

**Comment by depappas-sifive** [comment]
"There is no scalastyle-config.xml in rocket-chip."

Correct. But there is a  scalastyle-config.xml in federation which is used
by Intellij for both federation and rocket-chip.

**Comment by depappas-sifive** [comment]
The Described Reg annotation was removed. I did not create it.

File: src/main/scala/regmapper/Annotation.scala (hunk 6 above)
Comment (mwachs5) [reply src/main/scala/regmapper/Annotation.scala]: Not sure if you still have a pending change, but the DescribedReg annotation still exists in this version

Comment (hcook) [reply src/main/scala/regmapper/Annotation.scala]: I added it as an initial proposal for @depappas-sifive to work from, but I think we've decided to go the route of annotating the entire map at once inside the call to `regmap`.

I think the only thing we're giving up is the opportunity to annotate the specific `Reg` component with its description (rather than annotation the enclosing `Module` with the entire map). Annotating the `Reg` component in particular would make it much easier to extract e.g. the final signal name. Sadly, not all information is known inside of the call to `DescribedReg` (like the base address). A combination approach might be preferable but we can try that in a follow-up PR.

Comment (mwachs5) [reply src/main/scala/regmapper/Annotation.scala]: Yeah... but I expect there will be many more `RegFieldDesc` than there are `DescribedRegs`.

Comment (depappas-sifive) [reply src/main/scala/regmapper/Annotation.scala]: removed

Comment (depappas-sifive) [reply src/main/scala/regmapper/Annotation.scala]: Removed
